[David Liu] iP

Kangxun opened the pull request:
## Calico

> “An idea that is not dangerous is unworthy of being called an idea at all.” – Oscar Wilde ([source](https://www.goodreads.com/quotes/tag/ideas))

Calico _frees your mind_ of having to remember things you need to do. It's,
* text-based 
* easy to learn 👍🏻 
* ~~FAST~~ **SUPER FAST** to use 💯 

All you need to do is,

1. download it from [here](https://nus-cs2103-ay2122s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks. 
4. **let it** manage your tasks for you 😆 
    
And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines _(coming soon)_
- [ ] Reminders _(coming soon)_

-------------
If you Java programmer, you can use it to practice Java too. Here's a `main` method example:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Kangxun committed:
Redo Level 1. Greet, Echo, Exit

Kangxun committed:
Redo Level 2. Add, List

Kangxun committed:
Redo Level 3. Mark as Done

Kangxun committed:
Redo Level 4. ToDos, Events, Deadlines

Kangxun committed:
Added Automated Text UI Testing

Kangxun committed:
Redo Level 5. Handle Errors

Kangxun committed:
Redo Level 6. Delete

Kangxun committed:
Add Level-7 Save

Kangxun committed:
Add Level 8. Dates and Times

Kangxun committed:
Merge branch 'branch-Level-7'

Kangxun committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Duke.java
#	src/main/java/Event.java

Kangxun committed:
Make the code more OOP

Kangxun committed:
Divide classes into packages

Kangxun committed:
Add JUnit tests

Kangxun committed:
Add JavaDoc comments

Kangxun committed:
Tweak the code to comply with a coding standard

Kangxun committed:
Add Level 9. Find

Kangxun committed:
Merge branch 'branch-A-JavaDoc'

Kangxun committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/DukeException.java
#	src/main/java/duke/command/Todo.java

Kangxun committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/duke/task/Task.java
#	src/main/java/duke/task/TaskList.java
#	src/main/java/duke/util/Ui.java

nzixuan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,156 @@
+ */
+public class Duke {
+    private Storage storage;
+    private TaskList cmdList;
```
`cmdList` might be hard to pronounce and might not be the best choice of name according to the [Code Quality Guideline](https://nus-cs2103-ay2122s1.github.io/website/se-book-adapted/chapters/codeQuality.html). 
Perhaps changing to `commandList` would be clearer?

nzixuan commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,156 @@
+
+        while (Parser.isNotBye(cmd)) {
+            try {
+                if (Parser.isList(cmd)) {       // when command given is "list"
```
Should the comment be in a separate line from the code?
This is not explicitly stated in the Java coding standard but can help with consistency with the rest of the code.

Suggestion :
```java
if (Parser.isList(cmd)) {       
    // when command given is "list"
    ui.printList(cmdList);
```
I have noticed this issue a several times in this file as well.

nzixuan commented on src/main/java/duke/util/Parser.java:
```diff
@@ -0,0 +1,190 @@
+     * @param desc Command description to be processed.
+     * @return Deadline information.
+     */
+    public static String getDlInfo(String desc) {
```
`getDlInfor` Perhaps Dl might raise uncertainty for it's meaning in the method header. Spelling the full `getDeadlineInfo` might help to reduce uncertainty.

nzixuan reviewed (commented):
Overall good effort in keeping to many of the coding standard guidelines and best practices! Only minor commenting and naming errors are found throughout the files. Good job!

Kangxun commented:
thank you for the inputs @nzixuan 😄 i agree with you, ill do the necessary changes 👍

Kangxun committed:
Automate project builds using Gradle

Kangxun committed:
Merge branch 'branch-A-Gradle'

Kangxun committed:
Use CheckStyle

Kangxun committed:
Merge branch 'branch-A-CheckStyle'

Kangxun committed:
Add a GUI to Duke. Use the JavaFX technology to implement the GUI.

Kangxun committed:
Merge branch 'branch-Level-10'

adamozh commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,163 @@
+     * @param cmd Command from user
+     * @return Chatbot response
+     */
+    public String getResponse(String cmd) {
```
For better code readability, you could avoid arrowhead style code with deep nesting and long functions! (From week 5 topics) Perhaps you could abstract out some of the functionality into a different method to make it shorter.

adamozh commented on src/main/java/duke/command/Deadline.java:
```diff
@@ -0,0 +1,90 @@
+     *
+     * @return Deadline as a string.
+     */
+    @Override
```
I don't think JavaDocs are necessary for toString methods! (I hope I'm not wrong)

adamozh commented on src/main/java/duke/util/Storage.java:
```diff
@@ -0,0 +1,123 @@
+ * Allows for saving and loading tasks.
+ */
+public class Storage {
+    protected static File f;
```
Perhaps this file could have a more descriptive name?

adamozh reviewed (commented):
Looks good to me!

Smanmos reviewed (commented):
All looks good to me!

Kangxun committed:
Use Assertions

Important assumptions required for program to work are not checked for.

Documenting important assumptions that should hold at various points in the code helps with understanding of code and checks if program is running properly.

Let's,
* include assert statements to define assumptions about the program state
* enable assertions in Gradle run task

Integrating with Gradle allows for a smoother workflow in the development process.

Kangxun committed:
Improve code quality

Program code can be improved for readability.

Critically examining the code and refactoring helps make the code more readable and more easily understood.

Let's,
* avoid long methods
* avoid deep nesting
* make the happy path (the execution path taken when everything goes well) prominent
* extract methods

Refactoring code in such a way helps to make the program execution path and methods clearer to readers.

Kangxun commented:
thank you for the valuable inputs @moreTriangles @kawaiigabrielneko 😃 ill make the relevant changes 👌🏻

Kangxun committed:
Deal duplicate items

Duplicate tasks can be added onto the task list.

Handling duplicate items helps to keep the task list neat.

Let's,
* add the ability to recognize duplicate tasks
* deal with duplicate tasks when adding to task list

Preventing duplicate tasks from being added to task list keeps the task list tidy.

Kangxun committed:
Merge pull request #5 from Kangxun/branch-A-Assertions

Use Assertions

Kangxun committed:
Merge branch 'master' into branch-A-CodeQuality

# Conflicts:
#	src/main/java/duke/Duke.java

Kangxun committed:
Merge branch 'master' into branch-C-DetectDuplicates

Kangxun committed:
Merge pull request #2 from Kangxun/branch-A-CodeQuality

Improve code quality

Kangxun committed:
Merge branch 'master' of https://github.com/Kangxun/ip

Kangxun committed:
Merge branch 'master' into branch-C-DetectDuplicates

Kangxun committed:
Merge pull request #3 from Kangxun/branch-C-DetectDuplicates

Deal duplicate items

Kangxun committed:
Merge branch 'master' of https://github.com/Kangxun/ip

Kangxun committed:
Improve the GUI

Kangxun committed:
Rename chatbot

Kangxun committed:
Add a representative UI screenshot

Kangxun committed:
Add a User Guide

Kangxun committed:
Set theme jekyll-theme-midnight

Kangxun committed:
Set theme jekyll-theme-cayman

Kangxun committed:
Fix table layout in docs/README.md

Kangxun committed:
Merge remote-tracking branch 'origin/master'
